src/lightmap.cpp

```diff
@@ -186,6 +186,86 @@ void map::apply_character_light( player &p )
+// toward the lower limit. Since it's sunlight, the rays are parallel.
+// Each layer consults the next layer up to determine the intensity of the light that reaches it.
+// Once this is complete, additional operations add more dynamic lighting.
+void map::build_sunlight_cache( int zlev )
+{
+    level_cache &map_cache = get_cache( zlev );
+    auto &lm = map_cache.lm;
+    // Grab illumination at ground level.
+    const float outside_light_level = g->natural_light_level( 0 );
+    const float inside_light_level = ( outside_light_level > LIGHT_SOURCE_BRIGHT ) ?
+                                     LIGHT_AMBIENT_LOW + 1.0 : LIGHT_AMBIENT_MINIMAL;
+    // Handling when z-levels are disabled is based on whether a tile is considered "outside".
+    if( !zlevels ) {
+        const auto &outside_cache = map_cache.outside_cache;
+        for( int x = 0; x < MAPSIZE_X; x++ ) {
+            for( int y = 0; y < MAPSIZE_Y; y++ ) {
+                if( outside_cache[x][y] ) {
+                    lm[x][y].fill( outside_light_level );
+                } else {
+                    lm[x][y].fill( inside_light_level );
+                }
+            }
+        }
+        return;
+    }
+    // If uppermost level, just apply weather illumination since there's no opportunity
+    // for light to be blocked.
+    if( zlev == OVERMAP_HEIGHT ) {
+        for( int x = 0; x < MAPSIZE_X; x++ ) {
+            for( int y = 0; y < MAPSIZE_Y; y++ ) {
+                lm[x][y].fill( outside_light_level );
+            }
+        }
+        return;
+    }
+
+    // Replace this with a calculated shift based on time of day and date.
+    // At first compress the angle such that it takes no more than one tile of shift per level.
+    // To exceed that, we'll have to handle casting light from the side instead of the top.
+    point offset( 0, 0 );
+    const level_cache &prev_map_cache = get_cache_ref( zlev + 1 );
+    const auto &prev_lm = prev_map_cache.lm;
+    const auto &prev_transparency_cache = prev_map_cache.transparency_cache;
+    const auto &prev_floor_cache = prev_map_cache.floor_cache;
+    const auto &outside_cache = map_cache.outside_cache;
+    const float sight_penalty = weather_data( g->weather ).sight_penalty;
+    for( int x = 0, prev_x = offset.x; x < MAPSIZE_X; x++, prev_x++ ) {
+        bool x_inbounds = true;
+        if( prev_x < 0 || prev_x >= MAPSIZE_X ) {
+            x_inbounds = false;
+        }
+        for( int y = 0, prev_y = offset.y; y < MAPSIZE_Y; y++, prev_y++ ) {
+            bool inbounds = true;
+            if( !x_inbounds || prev_y < 0 || prev_y >= MAPSIZE_Y ) {
+                inbounds = false;
+            }
+            four_quadrants prev_light( outside_light_level );
+            float prev_transparency = static_cast<float>( LIGHT_TRANSPARENCY_OPEN_AIR );
+            if( inbounds ) {
+                prev_light = prev_lm[ prev_x ][ prev_y ];
```

Comment:
This code can be reached in the case that `x` is not in-bounds, in which case you're accessing outside the array.

Reply:
I'm not clear how that happens.
If prev_x is not inbounds, x_inbounds == false, which means inbounds is set to false on line 244.

I'd love a cleaner way to state this, but I don't see it.
Maybe I should just bite the bullet and do the whole inbounds check on each loop iteration instead of being parsimonious with checks.

Reply:
You're right, I misread.  Never mind.